File: src/csharp/VRE.Vridge.API.Client/Remotes/Beacons/VridgeServerBeaconList.cs
using System;
using System.Collections.Generic;
using System.Text;
using VRE.Vridge.API.Client.Messages.v3.Discovery;

namespace VRE.Vridge.API.Client.Remotes.Beacons
{
    class VridgeServerBeaconList
    {
        private readonly Dictionary<string, VridgeServerBeacon> timedList = new Dictionary<string, VridgeServerBeacon>();

        public void Add(Beacon beacon, string endpoint)
        {
            if (timedList.ContainsKey(endpoint))
            {
                timedList.Remove(endpoint);
            }

            timedList.Add(endpoint, new VridgeServerBeacon(beacon, endpoint));
        }

        public List<VridgeServerBeacon> FreshServers
        {
            get
            {
                var beacons = new List<VridgeServerBeacon>();
                foreach (var beacon in timedList.Values)
                {
                    if (beacon.IsFresh)
                    {
                        beacons.Add(beacon);
                    }
                }

                return beacons;
            }
        }
    }
}

File: src/java/VRE.Vridge.API/VRE.Vridge.API.Client/src/main/java/com/riftcat/vridge/api/client/java/remotes/beacons/VridgeServerBeaconList.java
package com.riftcat.vridge.api.client.java.remotes.beacons;

import com.riftcat.vridge.api.client.java.proto.Beacon;

import java.net.InetAddress;
import java.util.ArrayList;
import java.util.Dictionary;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

public class VridgeServerBeaconList {
    private HashMap<InetAddress, VridgeServerBeacon> timedList = new HashMap<InetAddress, VridgeServerBeacon>();

    public synchronized void add(Beacon beacon, InetAddress endpoint){
        if(timedList.containsKey(endpoint)){
            timedList.remove(endpoint);
        }

        timedList.put(endpoint, new VridgeServerBeacon(beacon, endpoint));
    }

    public synchronized List<VridgeServerBeacon> getFreshServers(){
        LinkedList<VridgeServerBeacon> beacons = new LinkedList<VridgeServerBeacon>();
        for (VridgeServerBeacon vridgeServerBeacon : timedList.values()) {
            if(vridgeServerBeacon.isFresh()){
                beacons.add(vridgeServerBeacon);
            }
        }

        return beacons;
    }
}
